File: Tutorials/ObjC/Example7.3/Example7.3/PlaylistViewController.m
//
//  PlaylistViewController.m
//  Example7
//
//  Created by Penthera on 3/3/21.
//

#import "PlaylistViewController.h"
#import <VirtuosoClientDownloadEngine/VirtuosoClientDownloadEngine.h>

@interface PlaylistViewController () <VirtuosoDownloadEngineNotificationsDelegate>

@property (nonatomic, strong)VirtuosoDownloadEngineNotificationManager* notificationManager;


@end

@implementation PlaylistViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    
    self.tableView.rowHeight = UITableViewAutomaticDimension;
    self.tableView.estimatedRowHeight = 80;

    self.notificationManager = [[VirtuosoDownloadEngineNotificationManager alloc]initWithDelegate:self];
    [self refreshView];
}

#pragma mark - Table view data source

- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView {
    return 1;
}

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
    return self.playlist.items.count;
}

-(UITableViewCell*)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath
{
    UITableViewCell* cell = [tableView dequeueReusableCellWithIdentifier:@"defaultCell" forIndexPath:indexPath];
    
    NSDateFormatter* formatter = [NSDateFormatter new];
    formatter.dateStyle = NSDateFormatterMediumStyle;
    formatter.timeStyle = NSDateFormatterMediumStyle;

    VirtuosoPlaylistItem* item = [self.playlist.items objectAtIndex:indexPath.row];
    cell.textLabel.text = item.assetID;
    NSMutableString* sb = [NSMutableString new];
    
    [sb appendFormat:@"Status: %@\n", [VirtuosoPlaylistItem statusAsString:item.itemState]];
    
    if (item.downloadComplete)
    {
        [sb appendFormat:@"Downloaded: yes %@\n", [formatter stringFromDate:item.downloadComplete]];
    }
    else
    {
        [sb appendString:@"Downloaded: no:\n"];
    }
    if (item.playbackDate)
    {
        [sb appendFormat:@"Playback: yes %@\n", [formatter stringFromDate:item.playbackDate]];
    }
    else
    {
        [sb appendString:@"Playback: no:\n"];
    }

    [sb appendFormat:@"Expired: %@\n", item.expired ? @"yes" : @"no"];
    [sb appendFormat:@"Pending Now: %@\n", item.pending ? @"yes" : @"no"];
    
    cell.detailTextLabel.text = sb;
    
    return cell;
}

-(void)addClicked
{
    VirtuosoPlaylist* playlist = self.playlist;
    
    if (!playlist) return;
    
    if (playlist.items.count >= 2 && playlist.items.count < 6)
    {
        NSString* episode = [NSString stringWithFormat:@"SERIES-8-EPISODE-%@", @(playlist.items.count + 1)];
        if ([playlist contains:episode])
        {
            return;
        }
        
        NSError* error = nil;
        [playlist append:[NSArray arrayWithObject:episode] error:&error];
        if (error)
        {
            VLog(kVL_LogError, "Error while appending to playlist: %@  error: %@",
                 playlist.name,
                 error.localizedDescription);
        }
        
        [self refreshView];
    }
}

-(void)refreshView
{
    dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_BACKGROUND, 0), ^{
        VirtuosoPlaylist*  playlist = [VirtuosoPlaylist find:self.playlist.name];
        dispatch_async(dispatch_get_main_queue(), ^{
            self.playlist = playlist;
            [self.tableView reloadData];

            if (self.playlist.items.count < 6)
            {
                self.navigationItem.rightBarButtonItem = [[UIBarButtonItem alloc]initWithTitle:@"Add" style:UIBarButtonItemStyleDone target:self action:@selector(addClicked)];
            }
            else
            {
                self.navigationItem.rightBarButtonItem = nil;
            }
        });
    });
}


- (void)downloadEngineDidFinishDownloadingAsset:(VirtuosoAsset * _Nonnull)asset {
    [self refreshView];
}

- (void)downloadEngineDidStartDownloadingAsset:(VirtuosoAsset * _Nonnull)asset {
    [self refreshView];
}

- (void)downloadEngineProgressUpdatedForAsset:(VirtuosoAsset * _Nonnull)asset {
    [self refreshView];
}

- (void)downloadEngineProgressUpdatedProcessingForAsset:(VirtuosoAsset * _Nonnull)asset {
    [self refreshView];
}

-(void)playlistChange:(VirtuosoPlaylist *)playlist
{
    [self refreshView];
}

@end

File: Tutorials/Swift/Example7.3/Example7.3/PlaylistViewController.swift
//
//  PlaylistViewController.swift
//  Example7
//
//  Created by Penthera on 7/10/19.
//  Copyright © 2019 Penthera. All rights reserved.
//

import UIKit

//
// This view demonstrates displaying

class PlaylistViewController: UITableViewController, VirtuosoDownloadEngineNotificationsDelegate {
    
    public var playlist: VirtuosoPlaylist?
    var notificationManager: VirtuosoDownloadEngineNotificationManager?

    // MARK: -
    // MARK: View Lifecycle

    override func viewDidLoad() {
        super.viewDidLoad()
        self.tableView.rowHeight = UITableView.automaticDimension
        self.tableView.estimatedRowHeight = 80
        
        self.notificationManager = VirtuosoDownloadEngineNotificationManager(delegate: self)
        guard let playlist = self.playlist else { return }
        self.navigationItem.title = playlist.name
    }
    
    override func viewWillAppear(_ animated: Bool) {
        super.viewWillAppear(animated)
        
        self.refreshView()
    }

    override func viewWillDisappear(_ animated: Bool) {
        super.viewWillDisappear(animated)
        self.notificationManager = nil;
    }
    
    // MARK: -
    // MARK: Table Support
    
    override func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        guard let playlist = self.playlist else { return 0 }
        return playlist.items.count
    }
    
    override func numberOfSections(in tableView: UITableView) -> Int {
        return 1
    }
    
    override func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        guard let cell = tableView.dequeueReusableCell(withIdentifier: "playlistItemCell") else {
            return UITableViewCell(style: .subtitle, reuseIdentifier: "playlistItemCell")
        }
        
        guard let playlist = self.playlist else {
            return cell
        }
        
        let playlistItem = playlist.items[indexPath.row]
        
        cell.textLabel?.text = playlistItem.assetID
        
        let dateformatter = DateFormatter()
        dateformatter.dateStyle = DateFormatter.Style.short
        dateformatter.timeStyle = DateFormatter.Style.short
        
        var detailText : String = ""
        detailText.append("Status: \(VirtuosoPlaylistItem.status(asString: playlistItem.itemState))\n")
        
        if nil != playlistItem.downloadComplete {
            detailText.append("Downloaded: yes \(dateformatter.string(from: playlistItem.downloadComplete!))\n")
        } else {
            detailText.append("Downloaded: no\n")
        }
        detailText.append("Deleted: \(playlistItem.userDeleted ? "yes" : "no")\n")
        
        if nil != playlistItem.playbackDate {
            detailText.append("Playback: yes \(dateformatter.string(from: playlistItem.playbackDate!))\n")
        } else {
            detailText.append("Playback: no\n")
        }
        
        detailText.append("Expired: \(playlistItem.expired ? "yes" : "no")\n")
        
        detailText.append("Pending Now: \(playlistItem.pending ? "yes" : "no")\n")
        if nil != playlistItem.pendingDate {
            detailText.append("Last Pending: \(dateformatter.string(from: playlistItem.pendingDate!))\n")
        }
        
        cell.detailTextLabel?.text = detailText;
        
        return cell
    }

    // MARK: -
    // MARK: Action handlers
    
    @objc func appendClicked()
    {
        guard let playlist = self.playlist else { return }
        
        // Append will add Episodes 4, 5, 6
        if (playlist.items.count >= 2 && playlist.items.count < 6)
        {
            let episode = "SERIES-8-EPISODE-\(playlist.items.count + 1)"
            
            if (playlist.contains(episode)) { return; }
            
            
            // Important:
            // When the following call completes, if the playlist had a pendingCount
            // we will immediately trigger an auto-download
            do {
            
                try playlist.append([episode])
                
            }
            catch let error as NSError
            {
                print("Playlist append failed with error: \(error.localizedDescription)")
            }
            
            self.refreshView()
        }
        else
        {
            self.navigationItem.rightBarButtonItem = nil
        }
    }
    
    // MARK: -
    // MARK: Internals
    func refreshView() {
        DispatchQueue.global(qos: .userInitiated).async
        {
            guard let playlist = self.playlist else { return }
            self.playlist = VirtuosoPlaylist.find(playlist.name)
            DispatchQueue.main.async
            {
                self.tableView.reloadData()
                
                guard let playlist = self.playlist else { return }
                
                if (playlist.items.count < 6) {
                    let append = UIBarButtonItem(title: "Append", style: .done, target: self, action: #selector(self.appendClicked))
                    self.navigationItem.rightBarButtonItem = append
                }
                else {
                    self.navigationItem.rightBarButtonItem = nil
                }
                

            }
        }
    }
    
    
    // MARK: -
    // MARK: VirtuosoDownloadEngineNotificationsDelegate support

    func downloadEngineDidStartDownloadingAsset(_ asset: VirtuosoAsset) {
        self.refreshView()
    }
    
    func downloadEngineProgressUpdated(for asset: VirtuosoAsset) {
        self.refreshView()
    }
    
    func downloadEngineProgressUpdatedProcessing(for asset: VirtuosoAsset) {
        self.refreshView()
    }
    
    func downloadEngineDidFinishDownloadingAsset(_ asset: VirtuosoAsset) {
        self.refreshView()
    }

    func downloadEngineInternalQueueUpdate() {
        self.refreshView()
    }
    
    func playlistChange(_ playlist: VirtuosoPlaylist) {
        self.refreshView()
    }
}
